mlflow/langchain/_langchain_autolog.py

```diff
@@ -180,6 +228,682 @@ def _chain_with_retriever(model):
+                        for k, v in doc.metadata.items()
+                    },
+                }
+                for doc in documents
+            ]
+            self._log_artifact(
+                {"documents": retriever_documents}, f"retriever_end_{retriever_span.span_id}"
+            )
+            self.metrics["step"] += 1
+            self.metrics["retriever_ends"] += 1
+            self.metrics["ends"] += 1
+
+        @override
+        def on_retriever_error(
+            self,
+            error: BaseException,
+            *,
+            run_id: UUID,
+            **kwargs: Any,
+        ):
+            """Run when Retriever errors."""
+            retriever_span = self._get_span_by_run_id(run_id)
+            retriever_span.add_event(
+                SpanEvent(
+                    "error",
+                    timestamp=int(datetime.now(timezone.utc).timestamp()),
+                    attributes={"error": str(error)},
+                )
+            )
+            error_message = self._get_stacktrace(error)
+            self._end_span(retriever_span, status=SpanStatus(TraceStatus.ERROR, error_message))
+            self.metrics["step"] += 1
+            self.metrics["errors"] += 1
+
+        @override
+        def on_agent_action(
+            self,
+            action: AgentAction,
+            *,
+            run_id: UUID,
+            parent_run_id: Optional[UUID] = None,
+            **kwargs: Any,
+        ) -> Any:
+            """Run on agent action."""
+            kwargs.update({"log": action.log})
+            agent_action_span = self._start_span(
+                span_name=action.tool,
+                parent_run_id=parent_run_id,
+                span_type=SpanType.AGENT,
+                run_id=run_id,
+                inputs={"tool_input": action.tool_input},
+                attributes=kwargs,
+            )
+            agent_action_span.add_event(
+                SpanEvent("start", timestamp=int(datetime.now(timezone.utc).timestamp()))
+            )
+            self._log_artifact(
+                {
+                    "tool": action.tool,
+                    "tool_input": action.tool_input,
+                    "log": action.log,
+                    "kwargs": kwargs,
+                },
+                f"agent_action_{agent_action_span.span_id}",
+            )
+            self.metrics["step"] += 1
+            self.metrics["agent_starts"] += 1
+            self.metrics["starts"] += 1
+
+        @override
+        def on_agent_finish(
+            self,
+            finish: AgentFinish,
+            *,
+            run_id: UUID,
+            parent_run_id: Optional[UUID] = None,
+            **kwargs: Any,
+        ) -> Any:
+            """Run on agent end."""
+            agent_span = self._get_span_by_run_id(run_id)
+            agent_span.add_event(
+                SpanEvent("end", timestamp=int(datetime.now(timezone.utc).timestamp()))
```

Comment:
```suggestion
                SpanEvent("end", timestamp=int(time.time()))
```

can we use `time.time()`?